Index: data/sm/sonic-sairedis/syncd/syncd.cpp
===================================================================
--- data.orig/sm/sonic-sairedis/syncd/syncd.cpp
+++ data/sm/sonic-sairedis/syncd/syncd.cpp
@@ -110,7 +110,9 @@ void notify_OA_about_syncd_exception()
 
             SWSS_LOG_NOTICE("sending switch_shutdown_request notification to OA");
 
-            notifications->send("switch_shutdown_request", "", entry);
+            std::string s = sai_serialize_switch_shutdown_request(gSwitchId);
+
+            notifications->send("switch_shutdown_request", s, entry);
 
             SWSS_LOG_NOTICE("notification send successfull");
         }
