Index: data/sm/sonic-swss/orchagent/vnetorch.cpp
===================================================================
--- data.orig/sm/sonic-swss/orchagent/vnetorch.cpp
+++ data/sm/sonic-swss/orchagent/vnetorch.cpp
@@ -89,7 +89,7 @@ bool VNetVrfObject::createObj(vector<sai
         sai_object_id_t router_id;
         if (vr_type != VR_TYPE::VR_INVALID && l_fn(router_id))
         {
-            SWSS_LOG_DEBUG("VNET vr_type %d router id %lx  ", vr_type, router_id);
+            SWSS_LOG_DEBUG("VNET vr_type %d router id %lx  ", static_cast<int>(vr_type), router_id);
             vr_ids_.insert(std::pair<VR_TYPE, sai_object_id_t>(vr_type, router_id));
         }
     }
Index: data/sm/sonic-swss/orchagent/vxlanorch.cpp
===================================================================
--- data.orig/sm/sonic-swss/orchagent/vxlanorch.cpp
+++ data/sm/sonic-swss/orchagent/vxlanorch.cpp
@@ -82,7 +82,7 @@ create_tunnel_map(MAP_T map_t)
 
     if (map_t == MAP_T::MAP_TO_INVALID)
     {
-        SWSS_LOG_ERROR("Invalid map type %d", map_t);
+        SWSS_LOG_ERROR("Invalid map type %d", static_cast<int>(map_t));
         return SAI_NULL_OBJECT_ID;
     }
