Index: usonic/sm/sonic-sairedis/vslib/src/sai_vs_interfacequery.cpp
===================================================================
--- usonic.orig/sm/sonic-sairedis/vslib/src/sai_vs_interfacequery.cpp
+++ usonic/sm/sonic-sairedis/vslib/src/sai_vs_interfacequery.cpp
@@ -511,6 +511,9 @@ bool check_ifname(
         if (c >= 'A' && c <= 'Z')
             continue;
 
+        if (c == ' ')
+            continue;
+
         SWSS_LOG_ERROR("invalid character '%c' in interface name %s", c, name.c_str());
         return false;
     }
